source/ai-server-logger.js

```diff
@@ -0,0 +1,122 @@
+                var delegateInfo = $delegate.info;
+                var delegateDebug = $delegate.debug;
+
+                var enhanceLoggingFunction = function(loggingFunction, severity) {
+                    return function() {
+                        var args = [].slice.call(arguments);
+                        loggingFunction.apply(null, args);
+
+                        var message = args.length === 1 ? _.first(args) : JSON.stringify(args);
+                        logger.send(message, severity);
+                    };
+                };
+
+                $delegate.error = enhanceLoggingFunction(delegateError, 'Error');
+                $delegate.warn = enhanceLoggingFunction(delegateWarn, 'Warning');
+                $delegate.info = enhanceLoggingFunction(delegateInfo, 'Info');
+                $delegate.debug = enhanceLoggingFunction(delegateDebug, 'Debug');
+                $delegate.trace = enhanceLoggingFunction(delegateDebug, 'Trace');
+
+                return $delegate;
+            });
+        })
+        .provider('AiServerLogger', function() {
+            var headers = {};
+            var serverPostEndPoint;
+            var queueSize = 100;
+            var postToServerDelay = 1000 * 60 * 10;  //10 minutes
+
+            var transformParameters = function() {
+                return {
+                    Message: arguments[0],
+                    Level: arguments[1]
+                };
+            };
+
+            var ServerLogger = (function() {
+                function ServerLogger(loggerConfig) {
+                    var config = loggerConfig;
+                    var queue = [];
+                    var timeOutId;
+                    var $http;
+
+                    this.send = function() {
+                        if(_.isUndefined(config.serverPostEndPoint)) { return; }
+                        var dto = transformParameters.apply(null, arguments);
+                        queue.push(dto);
+                        processQueue();
+                    };
+
+                    var processQueue = function() {
+                        if(angular.isDefined(timeOutId)) {
+                            if(queue.length >= config.queueSize) {
+                                clearTimeout(timeOutId);
+                                var payload = _.take(queue, config.queueSize);
+                                queue = _.slice(queue, config.queueSize);
+                                postToServer(payload);
+                            }
+                        }
+
+                        timeOutId = setTimeout(function() {
+                            timeOutId = undefined;
+                            if(queue.length === 0) { return ; }
+
+                            var payload = _.take(queue, config.queueSize);
+                            queue = _.slice(queue, config.queueSize);
+                            postToServer(payload);
+                        }, config.postToServerDelay);
+                    };
```

Comment:
I thought about debounce and decided I wanted a different behavior.  I want the POST to be fired as soon as we have queueSize messages even if we have not reached the timeout threshold.  The timeout threshold is there to ensure we regularly post things to the server so there is less chance of loosing information due to user closing the page/browser.

I will extract a function for repeated lines of code.